src/parser.rs

```diff
@@ -53,28 +79,34 @@ impl AST {
     pub fn root(&self) -> Root {
         Root::cast(self.node()).unwrap()
     }
-    /// Return all the errors that occured while parsing - NOT
-    /// including invalid stuff in the AST
-    pub fn root_errors(&self) -> &[ParseError] {
-        &self.errors
-    }
     /// Return all errors in the tree, if any
     pub fn errors(&self) -> Vec<ParseError> {
+        let ranges: HashSet<_> = self
+            .errors
+            .iter()
+            .filter_map(|e| match e {
+                ParseError::UnexpectedWanted(_, t, _) => Some(t),
+                ParseError::UnexpectedDoubleBind(t) => Some(t),
+                ParseError::UnexpectedExtra(t) => Some(t),
+                _ => None,
+            })
+            .collect();
         let mut errors = self.errors.clone();
```

Comment:
Am I correct that this patch supersedes root errors? Because if so, keeping all original errors here may result in duplicates?

Reply:
The error nodes that we collect from the tree should now be a superset of `self.errors`, basically containing the more detailed variant of the error node. This is why we filter `self.root().errors()` below. Ideally all of the errors in `self.root().errors()` should also be in self.errors, but right now there is no guarantee. Long story short: the filter below should exclude duplicates.